pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,430 @@
+	/// The day will be adjusted to the month.
+	///
+	/// NOTE: Only day 1 is supported by now
+	Monthly(u8),
+}
+
+/// Specify the paydown schedules of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PayDownSchedule {
+	/// The entire borrowed amount is expected to be paid back at the maturity
+	/// date
+	None,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct CashflowPayment<Balance> {
+	pub when: Seconds,
+	pub principal: Balance,
+	pub interest: Balance,
+}
+
+/// Specify the repayment schedule of the loan
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct RepaymentSchedule {
+	/// Expected repayments date for remaining debt
+	pub maturity: Maturity,
+
+	/// Period at which interest is paid
+	pub interest_payments: InterestPayments,
+
+	/// How much of the initially borrowed amount is paid back during interest
+	/// payments
+	pub pay_down_schedule: PayDownSchedule,
+}
+
+impl RepaymentSchedule {
+	pub fn is_valid(&self, now: Seconds) -> Result<bool, DispatchError> {
+		match self.interest_payments {
+			InterestPayments::None => (),
+			InterestPayments::Monthly(_) => {
+				let start = date::from_seconds(now)?;
+				let end = date::from_seconds(self.maturity.date())?;
+
+				// We want to avoid creating a loan with a cashflow consuming a lot of computing
+				// time Maximum 40 years, which means a cashflow list of 40 * 12 elements
+				if end.year() - start.year() > 40 {
+					return Ok(false);
+				}
+			}
+		}
+
+		Ok(self.maturity.is_valid(now))
+	}
+
+	pub fn generate_cashflows<Balance, Rate>(
+		&self,
+		origination_date: Seconds,
+		principal: Balance,
+		interest_rate: &InterestRate<Rate>,
+	) -> Result<Vec<CashflowPayment<Balance>>, DispatchError>
+	where
+		Balance: FixedPointOperand + EnsureAdd + EnsureDiv,
+		Rate: FixedPointNumber + EnsureDiv,
+	{
+		let start_date = date::from_seconds(origination_date)?;
+		let end_date = date::from_seconds(self.maturity.date())?;
+
+		let (timeflow, periods_per_year) = match &self.interest_payments {
+			InterestPayments::None => (vec![], 1),
+			InterestPayments::Monthly(reference_day) => (
+				date::monthly_intervals(start_date, end_date, (*reference_day).into())?,
+				12,
+			),
+		};
+
+		let total_weight = timeflow
+			.iter()
+			.map(|(_, weight)| weight)
+			.try_fold(0, |a, b| a.ensure_add(*b))?;
+
+		let interest_per_period = interest_rate
+			.per_year()
+			.ensure_div(Rate::saturating_from_integer(periods_per_year))?;
+
+		timeflow
+			.into_iter()
+			.map(|(date, weight)| {
+				let proportion = FixedU128::ensure_from_rational(weight, total_weight)?;
+				let principal = proportion.ensure_mul_int(principal)?;
+				let interest = interest_per_period.ensure_mul_int(principal)?;
+
+				Ok(CashflowPayment {
+					when: date::into_seconds(date)?,
+					principal,
+					interest,
+				})
+			})
+			.collect()
```

Comment:
We are creating diverging truths here. 
1. The interest of the expected cashflows is based on a monthly schedule, without any compounding. 
2. The interest of the loan itself is secondly compounding on the principal

I think we need to reiterate this and rather change the interest rate field on the loan itself and with it the `pallet-interest-accrual`

Also, what is proportion doing here? Why are we only calculating the interest based on a portion of the principal?

Reply:
Mmm... good point. Let me think about this!

Reply:
> Also, what is proportion doing here? Why are we only calculating the interest based on a portion of the principal?

Proportion is (or should be) the weight a month has regarding other months. This is because the borrowing could start in the middle of the month. In that case, such a month will have less weight than other entire months

Reply:
They are different use cases
- The interest on the loan itself is the amount that needs to be repaid now to close the loan
- The expected cashflows is for other use cases, including
  - The valuation being able to account for the expected cashflow dates (which we will adapt in the future with DCF)
  - The UI knowing when cashflows are expected
  - Enabling write-off policies to act on overdue cashflows

It also makes sense that the expected cashflows are non-compounding: they are assuming that each cashflow itself is paid in time and thus there is no compounding month over month.

Reply:
@hieronx and I just briefly synced and we definitely need to have cashflows and interest on the loan data being aligned. The easiest would be to make the cashflows compounding in the intervals they are expected. If they are repayed on time, the payments match. 


Due to secondly compounding this requirement will be impossible to be matched, so I think we need to take a deeper look here.

Reply:
> Due to secondly compounding this requirement will be impossible to be matched

This is because our monthly cashflow has day precision, right?

Reply:
It has monthly precisions IIUC